File: backend/app/api/analytics.py
"""
API роутер аналитики — дашборды для психологов и админов.
Эндпоинты под префиксом /api/analytics.
Все запросы требуют роль psychologist (или admin через require_role).
"""
from typing import Annotated

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models.user import User
from app.core.security import require_role
from app.services.analytics_service import (
    get_overview_stats,
    get_condition_distribution,
    get_category_stats,
    get_group_heatmap,
    get_trend_data,
    get_user_results,
    get_recent_results,
)

router = APIRouter()


@router.get("/overview")
async def overview(
    db: Annotated[AsyncSession, Depends(get_db)],
    _current_user: Annotated[User, Depends(require_role("psychologist"))],
) -> dict:
    """
    Общая статистика: пользователи, тесты, средний балл,
    тесты сегодня/за неделю.
    """
    return await get_overview_stats(db)


@router.get("/conditions")
async def conditions(
    db: Annotated[AsyncSession, Depends(get_db)],
    _current_user: Annotated[User, Depends(require_role("psychologist"))],
) -> list[dict]:
    """
    Распределение по уровням состояния — данные для PieChart.
    [{condition_level, count}]
    """
    return await get_condition_distribution(db)


@router.get("/categories")
async def categories(
    db: Annotated[AsyncSession, Depends(get_db)],
    _current_user: Annotated[User, Depends(require_role("psychologist"))],
) -> list[dict]:
    """
    Статистика по категориям — средний балл, кол-во тестов.
    Для BarChart сравнения.
    """
    return await get_category_stats(db)


@router.get("/heatmap")
async def heatmap(
    db: Annotated[AsyncSession, Depends(get_db)],
    _current_user: Annotated[User, Depends(require_role("psychologist"))],
) -> list[dict]:
    """
    Тепловая карта: средний балл по категориям × возрастным группам.
    """
    return await get_group_heatmap(db)


@router.get("/trends")
async def trends(
    db: Annotated[AsyncSession, Depends(get_db)],
    _current_user: Annotated[User, Depends(require_role("psychologist"))],
    days: int = Query(default=90, ge=7, le=365),
) -> list[dict]:
    """
    Тренд за последние N дней — ежедневный средний балл.
    Для LineChart.
    """
    return await get_trend_data(db, days)


@router.get("/user/{user_id}/results")
async def user_results(
    user_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    _current_user: Annotated[User, Depends(require_role("psychologist"))],
) -> list[dict]:
    """
    Детальные результаты конкретного пользователя — для PDF отчёта.
    """
    return await get_user_results(db, user_id)


@router.get("/recent")
async def recent(
    db: Annotated[AsyncSession, Depends(get_db)],
    _current_user: Annotated[User, Depends(require_role("psychologist"))],
    limit: int = Query(default=20, ge=1, le=100),
) -> list[dict]:
    """
    Последние N результатов тестов — с именами пользователей.
    Для таблицы в дашборде.
    """
    return await get_recent_results(db, limit)

File: frontend/src/api/analytics.ts
/**
 * API функции для аналитики — дашборды психолога и админа.
 * Все эндпоинты требуют роль psychologist или admin.
 */
import apiClient from './client'

/** Общая статистика */
export interface OverviewStats {
  total_users: number
  total_tests: number
  avg_score: number
  tests_today: number
  tests_this_week: number
}

/** Распределение по уровням состояния */
export interface ConditionItem {
  condition_level: string
  count: number
}

/** Статистика по категории */
export interface CategoryStat {
  category_id: number
  slug: string
  name_ru: string
  name_en: string
  name_kk: string
  avg_score: number
  test_count: number
}

/** Элемент тепловой карты */
export interface HeatmapItem {
  category_slug: string
  age_group: string
  avg_score: number
  count: number
}

/** Точка тренда */
export interface TrendPoint {
  date: string
  avg_score: number
  test_count: number
}

/** Результат пользователя (для PDF) */
export interface UserResult {
  id: number
  category_slug: string
  category_name_ru: string
  category_name_en: string
  raw_score: number
  condition_level: string
  ai_analysis: Record<string, unknown> | null
  created_at: string
}

/** Недавний результат (для таблицы админа) */
export interface RecentResult {
  id: number
  user_name: string
  user_email: string
  category_slug: string
  category_name_en: string
  raw_score: number
  condition_level: string
  created_at: string
}

/** Получить общую статистику */
export async function getOverview(): Promise<OverviewStats> {
  const response = await apiClient.get<OverviewStats>('/analytics/overview')
  return response.data
}

/** Получить распределение по уровням */
export async function getConditions(): Promise<ConditionItem[]> {
  const response = await apiClient.get<ConditionItem[]>('/analytics/conditions')
  return response.data
}

/** Получить статистику по категориям */
export async function getCategories(): Promise<CategoryStat[]> {
  const response = await apiClient.get<CategoryStat[]>('/analytics/categories')
  return response.data
}

/** Получить данные тепловой карты */
export async function getHeatmap(): Promise<HeatmapItem[]> {
  const response = await apiClient.get<HeatmapItem[]>('/analytics/heatmap')
  return response.data
}

/** Получить данные тренда за N дней */
export async function getTrends(days: number = 90): Promise<TrendPoint[]> {
  const response = await apiClient.get<TrendPoint[]>('/analytics/trends', {
    params: { days },
  })
  return response.data
}

/** Получить результаты конкретного пользователя */
export async function getUserResults(userId: number): Promise<UserResult[]> {
  const response = await apiClient.get<UserResult[]>(`/analytics/user/${userId}/results`)
  return response.data
}

/** Получить последние результаты тестов */
export async function getRecentResults(limit: number = 20): Promise<RecentResult[]> {
  const response = await apiClient.get<RecentResult[]>('/analytics/recent', {
    params: { limit },
  })
  return response.data
}
